### ISSUE TITLE
SSR my <app-root> loading doesn't appear in SSR build

### ISSUE THREAD

**Issue by lorenzochaudessolle** [angular/universal#1598]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/nguniversal

### Is this a regression?

 I never have test it in older version.

### Description

 Before i load my main component <app-root></app-root> (so my website), i 've wanted to show a spinner  like this :

![image](https://user-images.githubusercontent.com/24590805/78336733-2aade000-7590-11ea-91a8-8559c5b026ba.png)

But when i run my `npm run build:ssr`, the website load in white page and show directly the component, the spinner doesn't appear.

While in dev mode `ng serve`and in "standard build" `npm run build`  it works.

I put my code in index.html especially in <app-root> tags : 

`<app-root>
    <div class="app-loading">
      <div class="logo"></div>
      <svg class="spinner" viewBox="25 25 50 50">
        <circle class="path" cx="50" cy="50" r="20" fill="none" stroke-width="2" stroke-miterlimit="10" />
      </svg>
    </div>
  </app-root>`

the CSS is also put in the index.html

## 🔬 Minimal Reproduction

There is a little reproduction that I expect of my loading

https://stackblitz.com/edit/angular-issue-ssr-build-hyqpcq?embed=1&file=src/app/app.component.ts

## 🔥 Exception or Error

There is no error or Exception shown

## 🌍  Your Environment

**Angular Version:**
<pre><code>
                                       
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.0.2
Node: 12.14.0
OS: win32 x64

Angular: 9.0.1
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.2
@angular-devkit/build-angular     0.900.2
@angular-devkit/build-optimizer   0.900.2
@angular-devkit/build-webpack     0.900.2
@angular-devkit/core              9.0.2
@angular-devkit/schematics        9.0.2
@angular/cli                      9.0.2
@angular/platform-server          9.0.7
@ngtools/webpack                  9.0.2
@nguniversal/builders             9.0.2
@nguniversal/common               9.0.0
@nguniversal/express-engine       9.0.0
@schematics/angular               9.0.2
@schematics/update                0.900.2
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2

</code></pre>

**Anything else relevant?**

Thanks for your help :)

**Comment by lorenzochaudessolle** [comment]
> 
> 
> Maybe it is related to `app-shell` somehow

Thank you for your fast answser, i will check what is the `app-shell` !

**Comment by akires47** [comment]
@lorenzochaudessolle do u find a solution for this? 

I have the same problem with a loader, it works on `ng serve` but don't on `npm run build:ssr`

Thx!

**Comment by lorenzochaudessolle** [comment]
@akires47 Not yet..
Yes we have the same problem ! 
I continue my research :)

**Comment by lorenzochaudessolle** [comment]
After many researches, I did not find a mega awesome solution of the universe but a "workaround" that works.

### What is the problem ?

It's a fact, the code in the  `<app-root>` tags, in the index.html don't run in SSR build.

### How to fix

In **index.html**

First instead of placing the code in  `<app-root>` , place it in the ` <body>` tags,

like this : 

`<body>`
 `     <div id="loader">`
 `        <!-- Here is your code (spinner, text , etc...)   -->`
  `    </div>`
  `    <app-root>`
  `    </app-root>`
`</body>` 






Next step is to go in the component Page which will be loaded.
In my example, i have a basic routing infrastructure (`AppModule` and `app-routing.module` with **LazyLoading**)

So, to make it works on all my routes, I go in my `app.component`

In **app.component.html**

In my construtor I need : 
`constructor(@Inject(PLATFORM_ID) private platformId: Object, private renderer: Renderer2) { }`

`platformId` => To know if we are in front or in server part

`renderer` => We use Renderer2 because it's better than `ElementRef` to change the DOM
Why [here](https://medium.com/better-programming/angular-manipulate-properly-the-dom-with-renderer-16a756508cba)
And we use it to hide the loader manually.

Then, we need to use `AfterViewInit`, because it will be the event of the loading 's end.
We also use `isPlatformBrowser(this.platformId)` to create our condition to hide the loader only if we are on Browser (front part). Because the server part is call before the browser and it will be invisible for us .

So the code is : (sorry in advance for this indentation)

 `ngAfterViewInit(): void {`

    if (isPlatformBrowser(this.platformId)) {
      let loader = this.renderer.selectRootElement('#loader');
      if (loader.style.display != "none") loader.style.display = "none"; //hide loader
      console.log("test view init")
    }
 `}`

I put a conditional `display:none` because the `ngAfterViewInit` is called on each component call.

The loading animation begin after the css part and take end after all scripts are loaded.
![20200412_170044](https://user-images.githubusercontent.com/24590805/79072045-2fd10480-7cdf-11ea-9121-42050e655d62.gif)


In conclusion, i know this isn't the beautifulest solution but it run (in Angular 9).
If you have some question or notes to add, tell me 👍

**Comment by akires47** [comment]
@lorenzochaudessolle 

Thx for the solution it works perfect!

I looked the init of the app with SSR and i saw the render of the app is so quickly and the app-root always have something inside, because of the size of my app was imposible to research some solution, maybe i need to do more investigation about how SSR works and maybe we can find something "better" to use as a loder until response from SSR with the rendered app.

**Comment by lorenzochaudessolle** [comment]
@akires47 I am glad to have helped you ! Even if you do not have the perfect solution to your problem.
I'm in to find a better solution and i don't think we are alone in this situation. :) 

> because of the size of my app was imposible to research some solution, 

Your app is heavy or light ?

**Comment by akires47** [comment]
@lorenzochaudessolle is very light, i have 4 modules only, it's kind a static web page, i use always angular in everything, so i made it with angular too, but for the need of process open graph protocol for sharing in social media i put SSR, and the first thing i notice the loader is not there and was triying to find a solution, and i finded it here..

**Comment by scurk1415** [comment]
Isnt this a the normal behavior? You already have your page loaded when you use SSR so i don't think there should be a loader, cause the generated html gets inserted there

**Comment by lorenzochaudessolle** [comment]
I'm agreed with you @scurk1415 , so there isn't a "Good practice" 's solution to  make a loader ?
I thought a html generation which detect code in the `<app-root>` tags

**Comment by scurk1415** [comment]
Well you can do like this:

`@Directive({
	selector: '[appShellNoRender]'
})
export class AppShellNoRenderDirective implements OnInit {

	constructor(
		private viewContainer: ViewContainerRef,
		private templateRef: TemplateRef<any>,
		@Inject(PLATFORM_ID) private platformId) {}

	ngOnInit() {
		if (isPlatformServer(this.platformId)) {
			this.viewContainer.clear();
		}
		else {
			this.viewContainer.createEmbeddedView(this.templateRef);
		}
	}
}`

And in your app.component.html

`<ng-container *appShellNoRender>
	browser generate
</ng-container>
<ng-container *appShellRender>
	loader here
</ng-container>`

PS: appShellRender is the opposite of appShellNoRender

I found this solution somewhere (can't remember where, so i can't give credits to the original owner)

**Comment by akires47** [comment]
When the page is medium or heavy weight, there is a blank page moment until the page load is ready, is useful have a loader maybe in that moment and thats the idea from @lorenzochaudessolle

I saw what u say @scurk1415 when the page load instantly the app-root have html content inside, so for that reason the loader inside there is never show up

**Comment by CaerusKaru** [comment]
I think this is solved? To be honest I didn't really grok the issue to begin with, it looked like an edge case that has an edge solution. If anyone wants to clarify, we can maybe add docs on this case, otherwise I'm just going to close this issue as solved.
